### PR TITLE
Issue202 spectro x2d crystal

### PR DESCRIPTION
**Pull request by Didou09**
Motivation:
-------------
* This is a very large branch that was started several months ago to start implementing tools for D xray spectrometers
* It is not fully finished, but in its current version it is stable and unit tests are passing
* Instead of developping very large pieces of code on a signle branch, I think it is time to merge it into devel and try adressing each small change / new feature in a small dedicated branch, for cleaner coding, especially since we are now 2 working on the 2D X ray spectrometer

Main changes:
----------------
* a tofu.geom.CrystalBragg class is introduced, which contains method for computaing bragg diffraction, plotting, ray-tracing...
* Thanks to @adriendaros it also now contains methods for taking into account non-parallelism
* a tofu.spectro sub-module is introduced, which contains routines for fitting 1d (operational) and 2d (to be finished) spectra, as well as plotting tools
* Many units tests have been added, both for the Crystral class and for spectral fits.

Issues:
-------
* This PR addresses issue #202 and issue #494

Examples:
-----------

Plotting the crystal, detector and ray-tracing:
```
In [1]: import tofu as tf                                                                                                                                                              
/Home/DV226270/ToFu_All/tofu_git/tofu/tofu/imas2tofu/__init__.py:87: UserWarning: 
You do not seem to be using the latest IMAS version:
'module list' vs 'module av IMAS' suggests:
	- Current version: 3.30.0-4.8.4
	- Latest version : 3.32.1-4.9.1-R2019b
  warnings.warn(msg)

In [2]: cryst = tf.load('inputs_temp/TFG_CrystalBragg_ExpWEST_DgXICS_ArXVII_sh00000_Vers1.4.9-315-gc93fa7a3.npz')                                                                      
Loaded from:
    /Home/DV226270/ToFu_All/tofu_git/tofu/inputs_temp/TFG_CrystalBragg_ExpWEST_DgXICS_ArXVII_sh00000_Vers1.4.9-315-gc93fa7a3.npz

In [3]: det = dict(np.load('inputs_temp/det37_CTVD_incC4_New.npz', allow_pickle=True))                                                                                                 

In [4]: pts0, vect = cryst.get_rays_from_cryst(phi=np.pi, returnas='(pts, vect)')                                                                                                      

In [5]: pts = pts0 + np.r_[1.5, 3, 7]*vect[:, 0:1, 0]                                                                                                                                  

In [6]: dax = cryst.plot(pts=pts, det=det)                                                                                                                                             
/Home/DV226270/ToFu_All/tofu_git/tofu/tofu/geom/_comp_optics.py:669: RuntimeWarning: invalid value encountered in less
  if np.any(bragg < 0.):
/Home/DV226270/ToFu_All/tofu_git/tofu/tofu/geom/_core_optics.py:967: RuntimeWarning: invalid value encountered in greater_equal
  dk['det'][k>=0.] = k[k>=0.]
```

image

Fitting 1d spectra

```
In [34]: out = dict(np.load('inputs_temp/UV_spectra_sh55506.npz'))                                                                                                                     

In [35]: dlines = {'a{}'.format(ii): {'lambda0': vv} for ii, vv in enumerate(np.linspace(117.1e-10, 164.9e-10, 150))}                                                                  

In [36]: lk = sorted(dlines.keys())                                                                                                                                                    

In [37]: dconstraints = {'width': {'a': lk}, 'shift': {'a': lk}, 'double': False}                                                                                                      

In [38]: dconstants = {'shift': {'a': 0.}}                                                                                                                                             

In [39]: domain = {'lamb': [117e-10, 165e-10]}                                                                                                                                         

In [40]: dinput = tf.spectro.fit1d_dinput(lamb=out['lamb'], data=out['data'][:5, :], dlines=dlines, dconstraints=dconstraints, dconstants=dconstants, dscales=None, domain=domain)     

In [41]: tol = 1e-6                                                                                                                                                                    

In [42]: dfit1d = tf.spectro.fit1d(dinput=dinput, plot=False, chain=True, verbose=2, ftol=tol, xtol=tol, gtol=tol)                                                                     
spect      time (s)   cost       nfev       njev       msg       
---------- ---------- ---------- ---------- ---------- ----------
1 / 5      1.775379   7.011e+06  52         52         `ftol` termination condition is satisfied.
2 / 5      1.314889   1.083e+07  42         42         `ftol` termination condition is satisfied.
3 / 5      2.243551   9.372e+06  77         77         `ftol` termination condition is satisfied.
4 / 5      1.005408   8.598e+06  32         32         `ftol` termination condition is satisfied.
5 / 5      0.674391   9.591e+06  23         23         `ftol` termination condition is satisfied.
Total computation time:	7.015 s for 5 spectra (1.403 s per spectrum)

In [43]: dex = tf.spectro.fit1d_extract(dfit1d=dfit1d, pts_lamb_detail=True)                                                                                                           

In [44]: dax = tf.spectro._plot.plot_fit1d(dfit1d=dfit1d, dextract=dex, annotate=False, indspect=0)                
```        

image

### CODE REVIEW

File: tofu/spectro/_plot.py [hunk 1]
```diff
@@ -1188,7 +1216,8 @@ def plot_noise_analysis_scannbs(
         dax['conv'].legend(loc='best')
 
     # Particular cases
-    if dax.get('cases_fit') is not None and dnoise.get('nbsplines') is not None:
+    l0 = (dax.get('case_fit') is not None, dnoise.get('nbsplines') is not None)
+    if l0:
```
Comment (Didou09) [inline tofu/spectro/_plot.py:1220]: Beware, with your correction l0 is now a tuple, not a bool anymore!

**Review by Didou09 — COMMENTED** [review]
Good, except for l0 (see comment))

File: tofu/spectro/_plot.py [hunk 2]
```diff
@@ -1216,8 +1216,7 @@ def plot_noise_analysis_scannbs(
         dax['conv'].legend(loc='best')
 
     # Particular cases
-    l0 = (dax.get('case_fit') is not None, dnoise.get('nbsplines') is not None)
-    if l0:
+    if dax.get('case_fit') is not None and dnoise.get('nbsplines') is not None:
```
Comment (Didou09) [inline tofu/spectro/_plot.py:1219]: 👍

File: inputs_temp/XICS_allshots_C34.py [hunk 3]
```diff
@@ -1,9 +1,13 @@
+# -*- coding: utf-8 -*-
```
Comment (adriendaros) [inline inputs_temp/XICS_allshots_C34.py:1]: I've tried to find an error on a multi-line statement, like the failing result said but didn't find anything yet

Comment (Didou09) [reply inputs_temp/XICS_allshots_C34.py:1]: I've found it, can you try fixing the remaining non-confiormities (PEP8) in  tofu/spectro/_fit12d_funccostjac.py ?

Comment (adriendaros) [reply inputs_temp/XICS_allshots_C34.py:1]: I've already done it last thursday but i will redo

Comment (Didou09) [reply inputs_temp/XICS_allshots_C34.py:1]: (it was a hidden syntax error: a parenthesis was not closed)

**Comment by Didou09** [comment]
We can leave the last remaining PEP8 alerts in files:
image

There are only the ones in `tofu/spectro/_fit12d_funccostjac.py` left to fix and resolving the conflicts (I'll take care of conflicts).

**Comment by Didou09** [comment]
Well done with the PEP8 :+1:

**Pull request merged**